x-pack/platform/plugins/shared/task_manager/server/task_store.ts

```diff
@@ -156,6 +183,103 @@ export class TaskStore {
       maxRetries: 0,
     });
     this.requestTimeouts = opts.requestTimeouts;
+    this.security = opts.security;
+    this.spaces = opts.spaces;
+    this.canEncryptSavedObjects = opts.canEncryptSavedObjects;
+  }
+
+  public registerEncryptedSavedObjectsClient(client: EncryptedSavedObjectsClient) {
+    this.esoClient = client;
+  }
+
+  private canEncryptSo() {
+    return !!(this.esoClient && this.canEncryptSavedObjects);
+  }
+
+  private getSoClientForCreate(options: ApiKeyOptions) {
+    if (options.request) {
+      return this.savedObjectsService.getScopedClient(options.request, {
+        includedHiddenTypes: [TASK_SO_NAME],
+        excludedExtensions: [SECURITY_EXTENSION_ID, SPACES_EXTENSION_ID],
+      });
+    }
+    return this.savedObjectsRepository;
+  }
+
+  private async maybeGetApiKeyFromRequest(request?: KibanaRequest) {
+    if (!request) {
+      return null;
+    }
+
+    let userScopeAndApiKey;
+    try {
+      userScopeAndApiKey = await getApiKeyAndUserScope(
+        request,
+        this.canEncryptSo(),
+        this.security,
+        this.spaces
+      );
+    } catch (e) {
+      this.errors$.next(e);
+      throw e;
+    }
+
+    return userScopeAndApiKey;
+  }
+
+  private async bulkGetDecryptedTasks(ids: string[]) {
+    const result = new Map<string, ConcreteTaskInstance>();
+    if (!this.canEncryptSo() || !ids.length) {
+      return result;
+    }
+
+    const kueryNode = nodeBuilder.or(
+      ids.map((id) => {
+        return nodeBuilder.is(`${TASK_SO_NAME}.id`, `${TASK_SO_NAME}:${id}`);
+      })
+    );
+
+    const finder =
+      await this.esoClient!.createPointInTimeFinderDecryptedAsInternalUser<SerializedConcreteTaskInstance>(
+        {
+          type: TASK_SO_NAME,
+          filter: kueryNode,
+        }
+      );
+
+    for await (const response of finder.find()) {
+      response.saved_objects.forEach((savedObject) => {
+        result.set(savedObject.id, savedObjectToConcreteTaskInstance(savedObject));
+      });
+    }
+    await finder.close();
+
+    return result;
+  }
+
+  private async bulkGetAndMergeTasksWithDecryptedApiKey(tasks: ConcreteTaskInstance[]) {
+    const ids: string[] = [];
+
+    tasks.forEach((task) => {
+      if (task.apiKey) {
+        ids.push(task.id);
+      }
+    });
+
+    if (!ids.length) {
+      return tasks;
+    }
+
+    const decryptedTaskMap = await this.bulkGetDecryptedTasks(ids);
+
+    const tasksWithDecryptedApiKeys = tasks.map((task) => ({
+      ...task,
+      ...(decryptedTaskMap.get(task.id)?.apiKey
+        ? { apiKey: decryptedTaskMap.get(task.id)!.apiKey }
+        : {}),
```

Comment:
should we also return the `userScope` here?

Reply:
I think we should be ok here, since in line 276 the `userScope` should be included when the initial spread of `task` is done.